# EXP 2.1    
### digits 손글씨 이미지 데이터


In [19]:
# 필요한 모듈 import 하기
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# 데이터 준비
digits = load_digits()         


digits_data = digits.data             #feature data 지정
                                      #digits_data라는 변수에 digits의 데이터 입력

digits_label = digits.target          #label data 지정
                                      #digits의 정답 정보를 변수에 저장

print('"target name은 %s 입니다."' %digits.target_names)            #target names 출력  
print("\n")                                                       #digits의 종류 알 수 있다 (0-9까지의 숫자)


# train, test 데이터 분리 
# 모델 학습과 테스트용 문제지와 정답지 준비 
X_train, X_test, y_train, y_test = train_test_split (digits_data,      # 모델의 feature 
                                                    digits_label,      # 모델이 맞춰야 하는 정답 (0-9까지의 숫자)
                                                    test_size = 0.2,   # 전체 데이터 중 20%를 test데이터로 지정
                                                    random_state=7)    # 랜덤으로 train/test data 분리 


"target name은 [0 1 2 3 4 5 6 7 8 9] 입니다."




---
### 1) Decision Tree 모델

In [20]:
from sklearn.tree import DecisionTreeClassifier

#(5) 모델 학습 - X_train, y_train 활용하여 분류기 모델 만들기 / 성능 체크
decision_tree = DecisionTreeClassifier(random_state = 32)
decision_tree.fit(X_train,y_train)      # 모델학습 - training dataset에 맞게 패턴 파악 및 예측/학습
y_pred= decision_tree.predict(X_test)

print(classification_report(y_test,y_pred))

#성능평가
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test,y_pred)
accuracy

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.85       360



0.8555555555555555

---
### 2) Random Forest 모델

In [21]:
from sklearn.ensemble import RandomForestClassifier


random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))


from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test,y_pred)
accuracy


              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.93      1.00      0.97        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       0.93      1.00      0.96        37
           5       0.90      0.96      0.93        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       1.00      0.84      0.91        43
           9       0.94      0.94      0.94        32

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.97      0.96      0.96       360



0.9638888888888889

---
### 3) SVM 모델

In [22]:
from sklearn import svm
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test,y_pred)
accuracy


              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



0.9888888888888889

---
### 4) SGD Classifier 모델

In [23]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.90      0.86      0.88        42
           2       0.97      0.97      0.97        40
           3       0.73      0.97      0.84        34
           4       0.97      0.97      0.97        37
           5       0.88      1.00      0.93        28
           6       0.96      0.96      0.96        28
           7       0.91      0.97      0.94        33
           8       0.95      0.84      0.89        43
           9       1.00      0.69      0.81        32

    accuracy                           0.92       360
   macro avg       0.93      0.92      0.92       360
weighted avg       0.93      0.92      0.92       360



0.9222222222222223

---
### 5) Logistic Regression 모델

In [27]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()
# 오류나서 추가함 
logistic_model = LogisticRegression(solver='lbfgs', max_iter=10000).fit(X_train, y_train)
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)

accuracy


              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       1.00      1.00      1.00        37
           5       0.79      0.96      0.87        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       0.92      0.81      0.86        43
           9       0.97      0.88      0.92        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



0.95

---
# 분석하기 
정답(label)의 분포가 특별히 불균형 하지 않고 28-43으로 보인다. 따라서 **accuracy**로 모델의 성능을 평가 해도 좋다고 판단하였다.  

- 5가지 모델의 accuracy 비교 
  - Decision Tree   : 0.86  
  - Random Forest
   :     0.96
  - SVM
  : 0.99
  - SGD Classifier
   : 0.92
  - Logistic Regression
   : 0.95  
accuracy의 비교를 통하여 **SVM** 모델이 0.99로 가장 높게 나왔다.  
 
 